tempodb/backend/azure/azure.go

```diff
@@ -147,31 +154,26 @@ func (rw *readerWriter) Delete(ctx context.Context, name string, keypath backend
 func (rw *readerWriter) List(ctx context.Context, keypath backend.KeyPath) ([]string, error) {
 	keypath = backend.KeyPathWithPrefix(keypath, rw.cfg.Prefix)
 
-	marker := blob.Marker{}
 	prefix := path.Join(keypath...)
 
 	if len(prefix) > 0 {
 		prefix = prefix + dir
 	}
 
+	pager := rw.containerClient.NewListBlobsHierarchyPager(dir, &container.ListBlobsHierarchyOptions{
+		Include: container.ListBlobsInclude{},
+		Prefix:  &prefix,
+	})
+
 	objects := make([]string, 0)
-	for {
-		list, err := rw.containerURL.ListBlobsHierarchySegment(ctx, marker, dir, blob.ListBlobsSegmentOptions{
-			Prefix:  prefix,
-			Details: blob.BlobListingDetails{},
-		})
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
 		if err != nil {
 			return objects, errors.Wrap(err, "iterating tenants")
 		}
-		marker = list.NextMarker
-
-		for _, blob := range list.Segment.BlobPrefixes {
-			objects = append(objects, strings.TrimPrefix(strings.TrimSuffix(blob.Name, dir), prefix))
-		}
 
-		// Continue iterating if we are not done.
-		if !marker.NotDone() {
-			break
+		for _, b := range page.Segment.BlobPrefixes {
+			objects = append(objects, strings.TrimPrefix(strings.TrimSuffix(*b.Name, dir), prefix))
```

Comment:
i'm concerned about blindly derefencing b.Name. do we have any guarantees that this will always have a value? i'd rather return an error here than panic

Reply:
According to the SDK documentation, this value is required:
```go
type BlobPrefix struct {
	// REQUIRED
	Name *string `xml:"Name"`
}
```

The [API documentation](https://learn.microsoft.com/en-us/rest/api/storageservices/list-blobs) itself is unclear on whether this value is required.

I am fine with a defensive check.

Action taken: we now return an error if `b.Name == nil`.